clang/include/clang/Sema/Overload.h

```diff
@@ -1116,17 +1197,22 @@ class Sema;
   private:
     SmallVector<OverloadCandidate, 16> Candidates;
     llvm::SmallPtrSet<uintptr_t, 16> Functions;
+    SmallVector<DeferredTemplateOverloadCandidate, 8> DeferredCandidates;
```

Comment:
This is DEFINITELY cleaner, but a future 'optimization' here might store a single candidates list, then maintain a single 'index' to the first deferred candidate.  This would save a potential allocation/potentially improve perf, but at the expense of some complexity.  

I AM somewhat concerned about the increased size here though.  That SmallVector here is going to be HUGE, and Candidates is already quite large.  These are done often on the stack unfortunately, so this is going to have some effect on instantiation depth.

Reply:
I actually thought about that - but I have not implemented it yet.
Are the cases where we do a lot of overload resolution recursively that common?

Reply:
I recall Aaron I thought at one point working on reducing the overload-candidates size which showed great improvement, but that would have been ~9 months ago now.  I remember being surprised at how much of a difference it made.